Feature/login

deveroskp opened the pull request:
## 개요
<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
로그인 기능을 구현했습니다.

로그인 기능을 테스트하기 위해서, 로컬에서 터널링하여 연결하는 방식으로 임시로 db 연동했습니다.
테스트해보시려면
1. poetry로 패키지 설치하고
2.
.env.local 파일에서
DB_USER=
DB_PASSWORD=
DB_DATABASE=
에 개인 rds에  맞게 입력하고 alembic upgrade head 입력해서 마이그레이션 해주시면 됩니다.
3. ssh 명령어를 이용해서 터널을 생성합니다.
ssh -L {LOCAL_BIND_PORT}:{REMOTE_HOST}:{REMOTE_PORT} {ssh address} 

API 
1. /api/users/signup: 회원가입 -> username, login_id, login_password를 받아 회원가입을 진행하고 성공 시 'Success'를 반환합니다.
username과 login_id, login_password의 제한 조건은 fastapi 과제와 비슷한 형식으로 일단은 설정해놓았습니다.
2. /api/users/signin: 로그인 -> login_id와 login_password를 받아 로그인을 진행하고 성공 시 refresh_token을 쿠키에 저장하고 두 토큰의 값을 반환합니다.
3.  /api/users/me -> access_token을 헤더에 담아 요청하면 내 정보를 반환합니다.
4.  /api/users/refresh -> refresh_token을 쿠키에서 받아 access_token과 refresh_token을 갱신하고 반환합니다.
5. /api/users/logout -> refresh_token을 쿠키에서 받아 삭제하고 성공 시 'Success'를 반환합니다.

액세스 토큰는 헤더에 담고, 리프레시 토큰은 쿠키에 저장하는 방식으로 구현했는데, 찾아보니 쿠키에 저장하는 방식은 백엔드에서 처리하고, 헤더에 담는 작업은 프론트에서 처리하는 것 같습니다. 따라서 액서스 토큰을 단순히 반환만 했습니다.

참고자료 : https://bbogle2.tistory.com/entry/JWT-%EC%96%B4%EB%94%94%EC%97%90-%EC%A0%80%EC%9E%A5%ED%95%B4%EC%95%BC-%ED%95%A0%EA%B9%8C

<!---- 연관된 이슈가 있다면 태그주세요: ex) #(Issue Number)  -->
## PR 유형
어떤 변경 사항이 있나요?
- [✓ ] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [✓ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제
- [ ] 기타 : 
## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.  [Commit message convention](https://udacity.github.io/git-styleguide/) 참고
- [✓ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).
<!---- 노션 칸반보드 업데이트, 슬랙에 알리는 것도 잊지 마세요! -->
## 리뷰 요구사항 (선택)
<!---- 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요-->
많은 테스트케이스를 넣어본 건 아니라서, 확인해주시면 보완하겠습니다.

deveroskp committed:
로컬에서 db 연동 & jwt를 이용한 회원가입과 로그인 구현

deveroskp committed:
Chore: 토큰 저장 방식 변경

deveroskp committed:
Feat:로그아웃 기능 구현 & Fix:토큰 블락 코드 수정

deveroskp committed:
Feat: /me 엔드포인트 구현

deveroskp committed:
Fix: db version 파일 삭제

anandashin commented on .env.local:
```diff
@@ -0,0 +1,12 @@
+DB_PASSWORD=
+DB_DATABASE=
+
+JWT_SECRET_KEY=97ec0478040fc9b79be1e4dcdf9e1ba0d0a40372d20447b9ffed74a87f2c9b3d
```
.env 파일은 민감한 정보를 담고 있으니까 .gitignore에 넣어서 관리하는 게 좋을 것 같아요. (+ __pycache__ 도!)

anandashin commented on app/user/dto/responses.py:
```diff
@@ -0,0 +1,18 @@
+        return MyProfileResponse(
+            username=user.username, 
+            login_id=user.login_id, 
+            hashed_pwd=user.hashed_pwd
```
내 프로필 조회할 때 비밀번호는 안 보여주는게 좋을 것 같아요

anandashin commented on app/user/views.py:
```diff
@@ -0,0 +1,91 @@
+    user: User = Depends(login_with_header)
+):
+    return MyProfileResponse(username=user.username, login_id=user.login_id, hashed_pwd=user.hashed_pwd)
+
```
회원정보 수정하는 api도 추가하면 좋을 것 같아요! (username 변경 등)

anandashin commented on app/user/views.py:
```diff
@@ -0,0 +1,91 @@
+
+    return "Success"
+
+@user_router.post('/signin', 
```
로그인 요청을 보내면 터미널에 다음 오류가 찍히네요. 엑세스 토큰이랑 리프레시 토큰은 정상적으로 발급되긴 하는데..
(trapped) error reading bcrypt version
Traceback (most recent call last):
  File "/home/waffle/.cache/pypoetry/virtualenvs/22-5-team8-server-9fL7rEZa-py3.11/lib/python3.11/site-packages/passlib/handlers/bcrypt.py", line 620, in _load_backend_mixin
    version = _bcrypt.__about__.__version__
              ^^^^^^^^^^^^^^^^^
AttributeError: module 'bcrypt' has no attribute '__about__'

deveroskp replied:
bcrypt는 비밀번호 암호화할 때 사용되는 패키지인데, 결과상 암호화는 잘되는 것 같습니다. 찾아보니 bcrypt 최신버전에 오류가 있어 발생하는 것 같아 다운그레이드 해서 설치하니 해결됐습니다!

anandashin reviewed (commented):
postman 기준으로 다음 순서로 요청을 보냈을 때 에러가 발생한 경우가 있었습니다:

1. login -> logout (OK)
2. login -> me -> logout (Error raised)
3. login -> refresh (OK)
4. login -> me -> refresh (Error raise: Invalid Token)

에러 케이스에서 로그를 찍어보니 refresh, logout api의 
`refresh_token = request.cookies.get("refresh_token")`

refresh_token이 None으로 나오더라구요. login 요청 후 생성된 refresh token 쿠키가 다른 요청(/me, /signup ...)을 보내면 사라지는 것 같습니다.

minchok125 commented:
Swagger로 테스트를 해 보았는데, 상기해주신 cookie가 사라지는 문제가 발생하였다가 어느 시점부터는 다른 요청을 보내도 cookie가 사라지지 않고 정상적으로 작동하였습니다.

그래서 여러 번 배포해 보면서 cookie가 언제부터 유지되는 것인지 확인해 보았지만 특정한 조건을 찾지는 못했는데, 일단 배포 후 어느 시점부터는 정상적으로 작동하는 것으로 보아 테스트에는 큰 문제가 없을 것 같기는 합니다.

deveroskp commented:
쿠키가 없어지는 이유가 정말 사소한 이유였네요..
    response.set_cookie(
        key="refresh_token",
        value=refresh_token,
        httponly=True,
        expires=JWT_SETTINGS.refresh_token_expires_hours
        samesite="lax",
    )
 에서 expires에 들어가는 시간 단위가 초 단위였는데, JWT_SETTINGS.refresh_token_expires_hours에는 24로 설정되어있어서 24초만 유지되었던 것 같습니다.
 expires=JWT_SETTINGS.refresh_token_expires_hours*3600으로 수정했습니다.

deveroskp committed:
Feat: 유저 프로필 업데이트 기능 구현 & Fix: 리프레시 토큰 코드 수정
유저 프로필 업데이트 API를 구현했고, 리프레시 토큰의 쿠키 시간 설정이 잘못되어 있던 것을 수정했습니다. 그리고 내 프로필 조회할 때 비밀번호는 안 보여주게 수정했습니다.

minchok125 commented:
다시 테스트해 보았는데 쿠키가 사라지지 않고 잘 유지되는 것 확인했습니다!

anandashin commented on app/main.py:
```diff
@@ -4,11 +4,24 @@
 app.include_router(api_router, prefix="/api")
 
+origins = [
+    "*"
```
로컬환경 테스트에서는 상관없는데, 배포할 땐 전체허용보단 필요한 도메인(ex. 프론트엔드 도메인)만 명시해서 관리하면 더 좋을 것 같습니다!

deveroskp replied:
일단 후에 추가해보는 걸로 해보겠습니다!

anandashin reviewed (approved):
수고하셨습니다!

minchok125 reviewed (approved):
수고하셨습니다!!

deveroskp committed:
views.py에 response 예시 추가
